hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientWriteHandler.java

```diff
@@ -69,28 +66,15 @@ public void handle() throws Exception {
     }
 
     private void writeBuffer() throws IOException {
-        while (buffer.hasRemaining() && lastMessage != null) {
-            boolean complete = lastMessage.writeTo(buffer);
-            if (complete) {
+        while (lastMessage != null) {
+            socketChannel.write(lastMessage);
```

Comment:
I think that the performance drop may be caused due to change in number of socket write calls. In the new version, you make a call for each message. Let's assume that the number of bytes in the message is 1024 on average, then if buffer is 1024 K bytes, you increase the number of IO system call from 1024 K / 024 = 1024 instead of 1. So, we may need to reconsider this. May be this is a better choice to use an extra buffer :) We may also try this for C++ client as well.